numba_scipy/tests/test_special.py

```diff
@@ -61,6 +68,85 @@ def get_parametrize_arguments():
+    expected = ['double', 'double', 'double', 'double']
+    received = parse_capsule_name(input_capsule)
+    assert received == expected
+
+
+def test_parse_capsule_name_with_invalid_capsule():
+    with pytest.raises(ValueError) as excinfo:
+        input_capsule = '<TESTING object "double (double)" at 0x7f>'
+        parse_capsule_name(input_capsule)
+    assert "Unexpected capsule name" in str(excinfo.value)
+
+
+def test_parse_capsule_name_with_invalid_signature():
+    with pytest.raises(ValueError) as excinfo:
+        input_capsule = '<capsule object "TESTING" at 0x7f>'
+        parse_capsule_name(input_capsule)
+    assert "Unexpected signature" in str(excinfo.value)
+
+
+def test_de_mangle_function_name():
+    mangled_name = "__pyx_fuse_0pdtr"
+    received = de_mangle_function_name(mangled_name)
+    expected = "pdtr"
+    assert expected == received
+
+
+def test_de_mangle_function_name_with_invalid_name():
+    with pytest.raises(ValueError) as excinfo:
+        # The empty string was the only thing that the regex didn't recognise.
+        mangled_name = ""
+        print(de_mangle_function_name(mangled_name))
+    assert "Unexpected mangled name" in str(excinfo.value)
+
+
+@patch("numba_scipy.special.signatures.cysc")
+def test_get_signatures_from_pyx_capi(cysc_mock):
+    capsule = ('<capsule object "double (double, double, double, '
+               'int __pyx_skip_dispatch)" at 0x7f8c8d5f5150>')
+    mangled_name = "__pyx_fuse_0pdtr"
+    cysc_mock.__pyx_capi__ = {
+        mangled_name: capsule
+    }
+    expected = {
+        ('__pyx_fuse_0pdtr', float64, float64, float64, float64):
+        ('<capsule object "double (double, double, double, '
+         'int __pyx_skip_dispatch)" at 0x7f8c8d5f5150>')
+    }
+    received = get_signatures_from_pyx_capi()
+    assert expected == received
+
+
+@patch("numba_scipy.special.signatures.get_cython_function_address", Mock())
+@patch("numba_scipy.special.signatures.ctypes.CFUNCTYPE",
+       Mock(return_value=Mock(return_value='0123456789')))
+def test_generate_signatures_dicts():
+    signature_to_pointer = {
+        ('__pyx_fuse_0pdtr', float64, float64, float64, float64):
+        ('<capsule object "double (double, double, double, '
+         'int __pyx_skip_dispatch)" at 0x7f8c8d5f5150>')
+    }
+    expected = ({'pdtr': ((float64, float64, float64),)},
+                {('pdtr', float64, float64, float64): "0123456789"})
+    received = generate_signatures_dicts(signature_to_pointer)
+    assert expected == received
+
+
+def test_ensure_signatures_generated():
+    from numba_scipy.special.signatures import (name_to_numba_signatures,
+                                                name_and_types_to_pointer,
+                                                signature_to_pointer,
+                                                )
+    assert len(name_to_numba_signatures) != 0
+    assert len(name_and_types_to_pointer) != 0
+    assert len(signature_to_pointer) != 0
```

Comment:
should some of these containers also be the same size, if so, perhaps assert that?

Reply:
looks like it

```
In [3]: len(signatures.name_and_types_to_pointer)
Out[3]: 207

In [4]: len(signatures.name_to_numba_signatures)
Out[4]: 180

In [5]: len(signatures.signature_to_pointer)
Out[5]: 207
```

Reply:
changes add in 8cbba16

Reply:
thanks!